proposed/dynamic-component-loading.md

```diff
@@ -0,0 +1,110 @@
+# Dynamic component loading
+
+## Problem statement
+.NET Core runtime has only limited support for dynamically loading assemblies which have additional dependencies or possibly collide with the app in any way. Out of the box only these scenarios really work:
+* Assemblies which have no additional dependencies other than those found in the app itself (`Assembly.Load`)
+* Assemblies which have additional dependencies in the same folder and which don't collide with anything in the app itself (`Assembly.LoadFrom`)
+* Loading a single assembly in full isolation (from the app), but all its dependencies must come from the app (`Assembly.LoadFile`)
+
+Other scenarios are technically supported by implementing a custom `AssemblyLoadContext` but doing so is complex.  
+Additionally, there's no inherent synergy with the .NET Core SDK tooling. Components produced by the SDK can't be easily loaded at runtime.
+
+The goal of this feature is to provide an easy-to-use way to dynamically load a component with its dependencies.
+
+## Scenarios
+List of few scenarios where dynamic loading of full components is required:
+* MSBuild tasks - all tasks in MSBuild are dynamically loaded. Some tasks come with additional dependencies which can collide with each other or MSBuild itself as well.
+* Roslyn analyzers - similar to MSBuild tasks, the Roslyn compiler dynamically loads analyzers which are separate components with potentially conflicting dependencies.
+* XUnit loading tests - the test runner acts as an app and the test is loaded dynamically. The test can have any number of dependencies. Finding and resolving those dependencies is challenging.
+* ASP .NET's `dotnet watch` - ability to dynamically reload an app without restarting the process. Each version of the app is inherently in collision with any previous version. The old version should be unloaded.
```

Comment:
Please include ASP.NET Core plugins like z-pages (see [performance-profiling-controller.md](https://github.com/dotnet/designs/blob/master/accepted/performance-profiling-controller.md)) and monitoring tools like Application Insights.

Reply:
CC: @glennc @davidfowl

Reply:
@SergeyKanzhelev In which way do you think the performance profiling controller is related to the dynamic component loading? I can potentially see some use cases, but I don't see that as a good example of the scenarios we want to support (not saying it's not applicable).  

Monitoring tools and other profiler-like capabilities are probably more related to the [statup-hook] (https://github.com/dotnet/core-setup/blob/master/Documentation/design-docs/host-startup-hook.md) functionality. Also this new feature will explicitely not customize startup loading in any way. It will be on-demand - that is called by a user code. It is possible for startup-hooks to use this feature to load their own dependencies though.

I guess I'm missing the point here though - can you please describe in a bit more detail how do you think these are related?

Reply:
The discussion was how one can load ASP.NET components to implement profiling controller into an application that has no ASP.NET stack loaded or has it's own version of ASP.NET stack. Also those profiling controllers may implement some pluggable model to load data collection plug-ins on demand in runtime. Versioning and isolation was one of the discussion.

Reply:
For isolation, `AssemblyLoadContext` is definitely the solution. For the on demand runtime loading - that would be this feature (it's possible, but rather hard right now). Thanks for pointing that out - I'll include it in the list.